### PR TITLE
Feature based FTUE flows

### PR DESCRIPTION
**Pull request by ouchadam**
Depends on #4635 Part of #4584

- Creates a base for the iterative FTUE work (copied from `LoginActivity`) and allows it to co-exist with the Login2 (FTUE_WIP) 
- Ports the `LoginActivity2` to a `FTUEActivity` which dynamically routes to the iterative FTUE or FTUE_WIP variants
- Renames the `LoginVersion` to `LoginVariant` to allow for better naming
- Removes `SoftLogoutActivity2` as it wasn't exposed by the manifest and the implementation ended up using the legacy login flows 

My aim is to iteratively merge the changes from into FTUE_WIP into FTUE, eventually deleting the WIP

| FTUE | FTUE_WIP |
| --- | --- |
|ftue|ftue-wip

- Next steps, merge https://github.com/vector-im/element-android/pull/4590 into the `DefaultFTUEVariant`
- Add a way to access the features menu from the onboarding

### CODE REVIEW

File: vector/src/main/java/im/vector/app/features/ftue/DefaultFTUEVariant.kt [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+import im.vector.app.features.login.terms.LoginTermsFragmentArgument
+import im.vector.app.features.login.terms.toLocalizedLoginTerms
+import org.matrix.android.sdk.api.auth.registration.FlowResult
+import org.matrix.android.sdk.api.auth.registration.Stage
+import org.matrix.android.sdk.api.extensions.tryOrNull
+
+private const val FRAGMENT_REGISTRATION_STAGE_TAG = "FRAGMENT_REGISTRATION_STAGE_TAG"
+private const val FRAGMENT_LOGIN_TAG = "FRAGMENT_LOGIN_TAG"
+
+class DefaultFTUEVariant(
```
Comment (ouchadam) [inline vector/src/main/java/im/vector/app/features/ftue/DefaultFTUEVariant.kt:73]: this is a copy of https://github.com/vector-im/element-android/blob/develop/vector/src/main/java/im/vector/app/features/login/LoginActivity.kt

File: vector/src/main/java/im/vector/app/features/ftue/FTUEVariantFactory.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import im.vector.app.features.VectorFeatures
+import im.vector.app.features.login.LoginViewModel
+import im.vector.app.features.login2.LoginViewModel2
+import javax.inject.Inject
+
+class FTUEVariantFactory @Inject constructor(
+        private val vectorFeatures: VectorFeatures,
+) {
+
+    fun create(activity: FTUEActivity, loginViewModel: Lazy<LoginViewModel>, loginViewModel2: Lazy<LoginViewModel2>) = when (vectorFeatures.loginVariant()) {
```
Comment (ouchadam) [inline vector/src/main/java/im/vector/app/features/ftue/FTUEVariantFactory.kt:28]: these separate FTUE and FTUE_WIP variants only need to exist whilst the changes are ported over and updated from the WIP

File: vector/src/main/java/im/vector/app/core/platform/VectorBaseActivity.kt [hunk 3]
```diff
@@ -105,7 +105,7 @@ abstract class VectorBaseActivity<VB : ViewBinding> : AppCompatActivity(), Maver
     protected val viewModelProvider
         get() = ViewModelProvider(this, viewModelFactory)
 
-    protected fun <T : VectorViewEvents> VectorViewModel<*, *, T>.observeViewEvents(observer: (T) -> Unit) {
+    fun <T : VectorViewEvents> VectorViewModel<*, *, T>.observeViewEvents(observer: (T) -> Unit) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/platform/VectorBaseActivity.kt:108]: Why this change?

**Review by bmarty — COMMENTED** [review]
It sounds a bit over-engineered, but you seem to know what you are doing :)

File: vector/src/main/java/im/vector/app/core/platform/VectorBaseActivity.kt (hunk 3 above)
Comment (ouchadam) [reply vector/src/main/java/im/vector/app/core/platform/VectorBaseActivity.kt:108]: unfortunately the current visibility means we can only observe `ViewModel` `ViewEvents` from within an activity which is a bit too limiting, it's needed by https://github.com/vector-im/element-android/pull/4656/files#diff-98284f9358684c3751933502d694e1babd0a0b7038302282cd0f2037adb25970R109

I could copy the logic to avoid the visibility change but there's already some other public functions in the base activity so I thought this change wouldn't be _toooo_ out of place

Comment (bmarty) [reply vector/src/main/java/im/vector/app/core/platform/VectorBaseActivity.kt:108]: OK

**Review by ganfra — APPROVED** [review]
LGTM!
